### PR TITLE
Support packaging proto files into android archives(aar)

### LINKED ISSUE #56: proto import statement from Android project to Android project broken?
I accidentally opened this in the [old repository](https://github.com/aantono/gradle-plugin-protobuf/issues/25).

I have two Android libraries, both of which have protos. ProjectB depends on ProjectA -- compile project(':projectA'). The only non-standard thing I've done is to disable nano protos and use the normal style protos (reverting that doesn't solve my problem).

Both protos can compile just fine when they're independent. However, if I add an import statement within protoB.proto to protoA.proto, I get an error:
protoA.proto: File not found.
protoB.proto: Import "protoA.proto" was not found or had errors.

I think this issue is similar to [#22](https://github.com/aantono/gradle-plugin-protobuf/issues/22) but in my case, both projects are Android instead of plain Java. Should this work?

https://github.com/google/protobuf-gradle-plugin/blob/master/testProjectAndroid/build.gradle shows an Android app depending on protos defined in another Java project. If that works, shouldn't it still work when the other project is an Android project?

Should I move all protos into their own Java projects and have the Android projects depend on that instead?

Details:
classpath 'com.google.protobuf:protobuf-gradle-plugin:0.7.1'
proto_library_version = '3.0.0-beta-1'
google_proto_library = "com.google.protobuf:protobuf-java:${proto_library_version}"
google_protoc_artifact = "com.google.protobuf:protoc:${proto_library_version}"

---

I just confirmed my theory by converting ProjectA to a Java project (ProjectB is still an Android library). All of the sudden, protoB.proto is allowed to import a proto from protoA.proto.

In my case, I think I can just convert both of them to Java projects and be done with it. I made them Android library projects even though they only contain regular Java (there was some limitation with Android and eclipse long ago). But it still seems like this shouldn't be a restriction.

### PR DESCRIPTION
**Pull request by enbandari**
Support packaging proto files into android archives(aar) and extracting included proto files from android libraries. Tested on android gradle plugin v3.6.3. (#56)

### CODE REVIEW

**Comment by enbandari** [comment]
@googlebot I signed it!

**Comment by voidzcy** [comment]
Are you interested in adding tests for covering the usage of consuming an Android library project? (It might be non-trivial as we cannot reuse [testProjectBase](https://github.com/google/protobuf-gradle-plugin/tree/master/testProjectBase) to construct an Android Library project, would need to write one with the same proto sources but a build_base.gradle for android build)

**Comment by enbandari** [comment]
OK

--------------原始邮件--------------
发件人："Chengyuan Zhang "<notifications@github.com&gt;;
发送时间：2020年6月29日(星期一) 中午12:59
收件人："google/protobuf-gradle-plugin" <protobuf-gradle-plugin@noreply.github.com&gt;;
抄送："Bennyhuo "<bennyhuo@kotliner.cn&gt;;"Author "<author@noreply.github.com&gt;;
主题：Re: [google/protobuf-gradle-plugin] Fix #56. Support packaging proto files into android archives(aar). (#414)
-----------------------------------

 

 
Are you interested in adding tests for covering the usage of consuming an Android library project? (It might be non-trivial as we cannot reuse testProjectBase to construct an Android Library project, would need to write one with the same proto sources but a build_base.gradle for android build)

**Comment by enbandari** [comment]
I created a project [testProjectAndroidLibrary](https://github.com/google/protobuf-gradle-plugin/tree/77f2f27eb44ec25ef3fe55a750d7957573d047d1/testProjectAndroidLibrary) and simply renamed the [helloworld.proto](https://github.com/google/protobuf-gradle-plugin/blob/77f2f27eb44ec25ef3fe55a750d7957573d047d1/testProjectAndroidBase/src/main/proto/helloworld.proto) to [helloAar.proto](https://github.com/google/protobuf-gradle-plugin/blob/77f2f27eb44ec25ef3fe55a750d7957573d047d1/testProjectAndroidLibrary/src/main/proto/helloAar.proto) with the package name to 'io.grpc.helloaarexample'. And add the library as a dependency to [testProjectAndroidBase](https://github.com/google/protobuf-gradle-plugin/blob/77f2f27eb44ec25ef3fe55a750d7957573d047d1/testProjectAndroidBase/build_base.gradle#L113). Check [HelloAarActivity.java](https://github.com/google/protobuf-gradle-plugin/blob/77f2f27eb44ec25ef3fe55a750d7957573d047d1/testProjectAndroid/src/main/java/io/grpc/helloaarexample/HelloAarActivity.java).

BTW, there is no need to explicitly exclude proto files in packagingOptions. I have handled it in [ProtobufPlugin.groovy:237](https://github.com/google/protobuf-gradle-plugin/blob/77f2f27eb44ec25ef3fe55a750d7957573d047d1/src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy#L237).

**Comment by voidzcy** [comment]
> BTW, there is no need to explicitly exclude proto files in packagingOptions. I have handled it in [ProtobufPlugin.groovy:237](https://github.com/google/protobuf-gradle-plugin/blob/77f2f27eb44ec25ef3fe55a750d7957573d047d1/src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy#L237).

We want to only include proto files in packaging for android libraries, but not for applications. Applications are very sensitive to the apk size, including proto files in apk increases unnecessary bytes.

It looks the [default packagingOption](https://google.github.io/android-gradle-dsl/current/com.android.build.gradle.internal.dsl.PackagingOptions.html) does not include `**/*.proto`. If that true, we don't need that line.

**Comment by enbandari** [comment]
> > BTW, there is no need to explicitly exclude proto files in packagingOptions. I have handled it in [ProtobufPlugin.groovy:237](https://github.com/google/protobuf-gradle-plugin/blob/77f2f27eb44ec25ef3fe55a750d7957573d047d1/src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy#L237).
> 
> We want to only include proto files in packaging for android libraries, but not for applications. Applications are very sensitive to the apk size, including proto files in apk increases unnecessary bytes.
> 
> It looks the [default packagingOption](https://google.github.io/android-gradle-dsl/current/com.android.build.gradle.internal.dsl.PackagingOptions.html) does not include `**/*.proto`. If that true, we don't need that line.

Yes, this line just excludes proto files for android application. If it is a library project, don't exclude: 

```
        if (!project.android.hasProperty('libraryVariants')) {
          project.android.packagingOptions.exclude("**/*.proto")
        }
```

We should not include proto files for both apk size and safety reasons.

File: testProjectAndroidLibrary/src/main/proto/helloAar.proto [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// Message definitions to be used by integration test service definitions.
+syntax = "proto3";
+
+// From testProjectAndroidLibrary: src/main/proto
```
Comment (voidzcy) [inline testProjectAndroidLibrary/src/main/proto/helloAar.proto:34]: nit: this is from testProjectLite

File: testProjectAndroidLibrary/build.gradle [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+                // Options added to --grpc_out
+                option 'lite'
+            }
+        }
+    }
+}
+
+dependencies {
+    compile 'com.android.support:appcompat-v7:23.4.0'
+    compile 'com.squareup.okhttp:okhttp:2.7.5'
```
Comment (voidzcy) [inline testProjectAndroidLibrary/build.gradle:83]: Seems we do not need these dependencies here. It would be better to include only things we need.

**Comment by voidzcy** [comment]
Merging. Thanks for contributing.

**Pull request merged**